flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/storage/HashBufferAccumulatorTest.java

```diff
@@ -0,0 +1,151 @@
+import org.apache.flink.runtime.io.network.buffer.Buffer;
+import org.apache.flink.runtime.io.network.buffer.BufferPool;
+import org.apache.flink.runtime.io.network.buffer.NetworkBufferPool;
+import org.apache.flink.runtime.io.network.partition.hybrid.tiered.common.TieredStorageSubpartitionId;
+
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Collections;
+import java.util.Random;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/** Tests for {@link HashBufferAccumulator}. */
+class HashBufferAccumulatorTest {
+
+    public static final int NUM_TOTAL_BUFFERS = 1000;
+
+    public static final int NETWORK_BUFFER_SIZE = 1024;
+
+    private static final float NUM_BUFFERS_TRIGGER_FLUSH_RATIO = 0.6f;
+
+    private NetworkBufferPool globalPool;
+
+    @BeforeEach
+    void before() {
+        globalPool = new NetworkBufferPool(NUM_TOTAL_BUFFERS, NETWORK_BUFFER_SIZE);
+    }
+
+    @AfterEach
+    void after() {
+        globalPool.destroy();
+    }
+
+    @Test
+    void testAccumulateRecordsAndGenerateFinishedBuffers() throws IOException {
+        int numBuffers = 10;
+        int numRecords = 1000;
+        TieredStorageSubpartitionId subpartitionId = new TieredStorageSubpartitionId(0);
+        Random random = new Random();
+
+        TieredStorageMemoryManager tieredStorageMemoryManager =
+                createStorageMemoryManager(numBuffers);
+        HashBufferAccumulator bufferAccumulator =
+                new HashBufferAccumulator(NETWORK_BUFFER_SIZE, tieredStorageMemoryManager);
+
+        AtomicInteger numReceivedFinishedBuffer = new AtomicInteger(0);
+        bufferAccumulator.setup(
+                1,
+                ((subpartition, buffers) ->
+                        buffers.forEach(
+                                buffer -> {
+                                    numReceivedFinishedBuffer.incrementAndGet();
+                                    buffer.recycleBuffer();
+                                })));
+
+        int numSentBytes = 0;
+        for (int i = 0; i < numRecords; i++) {
+            int numBytes = random.nextInt(2 * NETWORK_BUFFER_SIZE) + 1;
+            numSentBytes += numBytes;
+            ByteBuffer record = generateRandomData(numBytes, random);
+            bufferAccumulator.receive(record, subpartitionId, Buffer.DataType.DATA_BUFFER);
+        }
+        ByteBuffer endEvent = EventSerializer.toSerializedEvent(EndOfPartitionEvent.INSTANCE);
+        bufferAccumulator.receive(endEvent, subpartitionId, Buffer.DataType.EVENT_BUFFER);
+
+        int numExpectBuffers =
+                numSentBytes / NETWORK_BUFFER_SIZE
+                        + (numSentBytes % NETWORK_BUFFER_SIZE == 0 ? 0 : 1);
+
+        assertThat(numReceivedFinishedBuffer.get()).isEqualTo(numExpectBuffers + 1);
```

Comment:
We should also test that event buffer is not requested from memory manager. Maybe introduce a new test method would be more suitable.

Reply:
The event buffer is indeed not requested from the memory manager (Events should use heap buffer).

But I Added a new test to test that the buffer should be finished when it is full, even though not receiving an event.